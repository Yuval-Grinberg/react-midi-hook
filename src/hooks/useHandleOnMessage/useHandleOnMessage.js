import COMMANDS from '../../constants/commands';

import { useCallback, useState } from 'react';
import { getNoteLetter, getMIDICommand } from '../../utils/conversions';

/**
 * Handles MIDI message events
 */
export default function useHandleOnMessage() {
  const [pressedKeys, setPressedKey] = useState([]);

  /**
   * Handles "NOTE_ON" events
   *
   * @param {Number} position the real key position
   * @param {Number} velocity the input velocity
   */
  const noteOn = useCallback(
    ({ position, velocity }) => {
      const octave = parseInt(position / 12 + 1, 10);
      const letter = getNoteLetter(position % 12);

      return setPressedKey((keysStillPressed) => [
        ...keysStillPressed,
        { position, octave, letter, velocity }
      ]);
    },
    [setPressedKey]
  );

  /**
   * Handles "NOTE_OFF" events
   *
   * @param {Number} position the real key position
   */
  const noteOff = useCallback(
    ({ position }) => {
      return setPressedKey((keysStillPressed) => [
        ...keysStillPressed.filter((key) => key.position !== position)
      ]);
    },
    [setPressedKey]
  );

  /**
   * Handles incoming MIDI events from input devices
   *
   * @param {Object} event
   */
  const onMessage = useCallback(
    (event) => {
      const [command, note, velocity] = event.data;

      // Real keyboard notes are different to MIDI notes
      const position = note - 21;
      const type = getMIDICommand(command);

      if (type === COMMANDS.NOTE_ON) return noteOn({ position, velocity });
      if (type === COMMANDS.NOTE_OFF) return noteOff({ position });
    },
    [noteOn, noteOff]
  );

  return { pressedKeys, onMessage };
}
