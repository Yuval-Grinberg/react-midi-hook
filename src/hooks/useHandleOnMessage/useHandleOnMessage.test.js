import { renderHook, act } from '@testing-library/react-hooks';
import useHandleOnMessage from './useHandleOnMessage';

describe('useHandleOnMessage', () => {
  describe('noteOn', () => {
    it('adds a note to the state when pressed', () => {
      const { result } = renderHook(() => useHandleOnMessage());

      act(() => {
        result.current.onMessage({
          data: [144, 22, 100]
        });
      });

      expect(result.current.pressedKeys).toMatchSnapshot();
    });
  });

  describe('noteOff', () => {
    it('removes the note from the state when pressed', () => {
      const { result } = renderHook(() => useHandleOnMessage());

      act(() => {
        result.current.onMessage({
          data: [144, 22, 100]
        });

        result.current.onMessage({
          data: [128, 22, 0]
        });
      });

      expect(result.current.pressedKeys).toMatchSnapshot();
    });
  });
});
