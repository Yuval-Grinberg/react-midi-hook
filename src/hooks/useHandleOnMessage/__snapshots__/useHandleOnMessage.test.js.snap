// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useHandleOnMessage noteOff removes the note from the state when pressed 1`] = `Array []`;

exports[`useHandleOnMessage noteOn adds a note to the state when pressed 1`] = `
Array [
  Object {
    "letter": "A#",
    "octave": 1,
    "position": 1,
    "velocity": 100,
  },
]
`;
